### PR TITLE
258 expose a transition path for solvers

### LINKED ISSUE #258: Expose a transition path for solvers
Initially, for demo-ing what that looks like.

### PR DESCRIPTION
**Pull request by alberto-scolari**
Adding interfaces to expose solvers on ALP transition path:
- (pre-existing) solver interface `solve_cg_*`
- (newly added) KML interface `KmlIss*` https://support.huawei.com/enterprise/en/doc/EDOC1100283141/eee6b84d/single-node-mode#EN-US_TOPIC_0000001468009088

### CODE REVIEW

File: bootstrap.sh [hunk 1]
```diff
@@ -84,6 +84,7 @@ the location where LPF is installed"
 	echo "  --generator=<value>                 - set the generator for CMake (otherwise use CMake's default)"
 	echo "  --show                              - show generation commands instead of running them"
 	echo "  --delete-files                      - delete files in the current directory without asking for confirmation"
+	echo "  --spblas-prefix                     - delete files in the current directory without asking for confirmation"
```
Comment (anyzelman) [inline bootstrap.sh]: Explantion is copied from delete-files instead of describing what --spblas-prefix does

File: include/graphblas/algorithms/conjugate_gradient.hpp [hunk 2]
```diff
@@ -154,19 +156,20 @@ namespace grb {
 				grb::identities::zero, grb::identities::one
 			>,
 			class Minus = operators::subtract< IOType >,
-			class Divide = operators::divide< IOType >
+			class Divide = operators::divide< IOType >,
+			typename RSI, typename NZI, Backend backend
```
Comment (anyzelman) [inline include/graphblas/algorithms/conjugate_gradient.hpp:159]: Why would we want to expose RSI, NZI, and backend on the algorithm level? (We never do that right?)

File: include/graphblas/algorithms/conjugate_gradient.hpp [hunk 3]
```diff
@@ -310,23 +313,6 @@ namespace grb {
...
-				ret = ret ? ret : grb::dot< descr_dense >( bnorm, temp, b, ring );
-			} else {
-				ret = ret ? ret : grb::dot< descr_dense >( bnorm, b, b, ring );
-			}
-			assert( ret == SUCCESS );
-
-			if( ret == SUCCESS ) {
-				tol *= sqrt( grb::utils::is_complex< IOType >::modulus( bnorm ) );
-			}
-
```
Comment (anyzelman) [inline include/graphblas/algorithms/conjugate_gradient.hpp]: The standard is to pass a relative tolerance to a solver, see e.g. https://docs.octave.org/latest/Specialized-Solvers.html. Two options: either roll this back, or allow for both variants on the ALP/GraphBLAS level (with the relative tolerance as default)

File: include/graphblas/algorithms/conjugate_gradient.hpp [hunk 4]
```diff
@@ -415,13 +401,6 @@ namespace grb {
 			// output that is independent of error code
 			iterations = iter;
 
-			// return correct error code
-			if( ret == SUCCESS ) {
-				if( sqrt( residual ) >= tol ) {
-					// did not converge within iterations
-					return FAILED;
-				}
-			}
```
Comment (anyzelman) [inline include/graphblas/algorithms/conjugate_gradient.hpp]: Even if using an absolute norm the algorithm should return FAILED if it does not reach that tolerance

File: include/graphblas/nonblocking/vector.hpp [hunk 5]
```diff
@@ -197,12 +197,44 @@ namespace grb {
 
 		friend class PinnedVector< D, nonblocking >;
```
Comment (anyzelman) [inline include/graphblas/nonblocking/vector.hpp:199]: Maybe add a comment here that these are native interface friends? (See e.g. nonblocking/matrix.hpp)

File: include/transition/kml_iss.h [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+int KmlIssCgCleanSI(KmlSolverTask **);
+int KmlIssCgCleanDI(KmlSolverTask **);
+
+
+#ifdef __cplusplus
+} // end extern "C"
+#endif
+
+#endif // end ifdef _H_ALP_KML_ISS
```
Comment (anyzelman) [inline include/transition/kml_iss.h:145]: Missing empty line below here

File: include/transition/kml_iss.h [hunk 7]
```diff
@@ -0,0 +1,129 @@
+
+
```
Comment (anyzelman) [inline include/transition/kml_iss.h]: One too many empty lines

File: include/transition/kml_iss.h [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+#include <stddef.h> // for size_t
+
+#include "solver.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
```
Comment (anyzelman) [inline include/transition/kml_iss.h]: Inconsistent use of two empty lines and one empty line for vertical spacing throughout the file, starting here -- please unify

File: src/transition/CMakeLists.txt [hunk 9]
```diff
@@ -18,48 +18,70 @@
...
+	add_transition_library( spsolver_nonblocking_static STATIC "spsolver_nonblocking" ${CMAKE_CURRENT_SOURCE_DIR}/solver.cpp
+		PUBLIC_LINK_LIBRARIES backend_nonblocking
+	)
 
-	install( TARGETS sparseblas_omp_static
-		EXPORT GraphBLASTargets
-		ARCHIVE DESTINATION "${SHMEM_BACKEND_INSTALL_DIR}"
+	add_transition_library( kml_solver STATIC "kml_solver" ${CMAKE_CURRENT_SOURCE_DIR}/kml_iss.cpp
+		PUBLIC_LINK_LIBRARIES spsolver_nonblocking_static
 	)
```
Comment (anyzelman) [inline src/transition/CMakeLists.txt]: I prefer that we by default only use the nonblocking backend to generate the solver library. Compiling it with OMP of course works but we want people to use the nonblocking backend, no?

Optionally, the emitting the libraries compiled with the blocking OpenMP backend can be enabled on request.

On the naming: the nonblocking solver libraries should just be called shmem or shmem_parallel, and should not be called `nonblocking`. Otherwise it is not immediately clear to users what it is for-- by contrast, users will always understand what is sequential or shmem parallel.
The blocking backend could be called shmem_blocking: since it would only be generated when the user asks for it, that user should know what is blocking and nonblocking, so we can use that nomenclature.

File: src/transition/CMakeLists.txt [hunk 10]
```diff
@@ -18,48 +18,70 @@
...
+	)
 
-	install( TARGETS sparseblas_static
-		EXPORT GraphBLASTargets
-		ARCHIVE DESTINATION "${SHMEM_BACKEND_INSTALL_DIR}"
+	set( _SPBLAS_PREFIX "kml_" )
+	if( SPBLAS_PREFIX )
+		set( _SPBLAS_PREFIX "${SPBLAS_PREFIX}_" )
+	endif()
+	add_transition_library( ${_SPBLAS_PREFIX}sparseblas_static STATIC "${_SPBLAS_PREFIX}sparseblas"
```
Comment (anyzelman) [inline src/transition/CMakeLists.txt]: Should the name not be something like `sparseblas_sequential` rather than "sparseblas"?

File: src/transition/kml_iss.cpp [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+			break;
+		}
+	}
+	return result;
+}
+
+#define KML_CG_PREFIXED( name ) KmlIssCg ## name
+
+
+int KML_CG_PREFIXED(InitSI)( KmlSolverTask **handle, int n, const float *a, const int *ja, const int *ia ) {
```
Comment (anyzelman) [inline src/transition/kml_iss.cpp]: Please check for line overflow (80 chars max, starting from first non-space char)

File: src/transition/kml_iss.cpp [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+	return sparse_err_t_2_int( err );
+}
+
+int KML_CG_PREFIXED(InitDI)( KmlSolverTask **handle, int n, const double *a, const int *ja, const int *ia ) {
+	sparse_err_t err = sparse_cg_init_dii( handle, n, a, ja, ia );
+	return sparse_err_t_2_int( err );
+}
+
+
+
```
Comment (anyzelman) [inline src/transition/kml_iss.cpp]: Remove superfluous empty lines (also below)

File: src/transition/kml_iss.cpp [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+		if( r != NO_ERROR ) {									\
+			err = ( r == NULL_ARGUMENT ) ?						\
+				KMLSS_NULL_ARGUMENT : KMLSS_INTERNAL_ERROR;		\
+		}														\
+		break;													\
+	}															\
+	default:													\
+		err = KMLSS_BAD_SELECTOR;								\
+	}															\
+	return err;
```
Comment (anyzelman) [inline src/transition/kml_iss.cpp]: The back-ticks are not space-aligned. Also, not better to define all macros somewhere at the top of this file?

File: src/transition/kml_iss.cpp [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+#define KMLSS_SOLVE( nb, sparse_err_t_val )				\
+	if( nb != 1 ) { return KMLSS_BAD_NB; }				\
+	int err = KMLSS_NO_ERROR;							\
+	const sparse_err_t r = sparse_err_t_val;			\
+	if( r != NO_ERROR ) {								\
+		err = ( r == NULL_ARGUMENT ) ?					\
+			KMLSS_NULL_ARGUMENT : KMLSS_INTERNAL_ERROR;	\
+	}													\
+	return err;
```
Comment (anyzelman) [inline src/transition/kml_iss.cpp]: same here

Comment (anyzelman) [reply src/transition/kml_iss.cpp]: (and further down this file also)

File: src/transition/solver.cpp [hunk 15]
```diff
@@ -0,0 +1,686 @@
...
+
+#include <graphblas/algorithms/conjugate_gradient.hpp>
+
+
+template< typename T, typename NZI, typename RSI >
+class CG_Data {
+
+	private:
+
+		typedef grb::Matrix< T, grb::config::default_backend, NZI, NZI, RSI > Matrix;
```
Comment (anyzelman) [inline src/transition/solver.cpp]: line overflow?

File: tests/smoke/conjugate_gradient.cpp [hunk 16]
```diff
@@ -136,6 +136,39 @@ void ioProgram( const struct input &data_in, bool &success ) {
...
+	grb::operators::add< ScalarType >, grb::operators::mul< ScalarType >,
+	grb::identities::zero, grb::identities::one
+>;
+
+template<
+	Backend backend
+> RC computeNorm(
+	const grb::Vector< ScalarType, backend > &v,
+	grb::Vector< ScalarType, backend > &temp,
+	BaseScalarType &norm
```
Comment (anyzelman) [inline tests/smoke/conjugate_gradient.cpp]: The code standard has that the output should be the first argument, not the last

File: tests/smoke/conjugate_gradient.cpp [hunk 17]
```diff
@@ -136,6 +136,39 @@ void ioProgram( const struct input &data_in, bool &success ) {
...
+			}, temp
+		);
+		rc = rc ? rc : grb::dot< descriptors::dense >( bnorm, temp, v, ring );
+	} else {
+		rc = rc ? rc : grb::dot< descriptors::dense >( bnorm, v, v, ring );
+	}
+	assert( rc == SUCCESS );
+
+	norm = rc ? ring.template getZero< BaseScalarType >() :
+		sqrt( grb::utils::is_complex< ScalarType >::modulus( bnorm ) );
```
Comment (anyzelman) [inline tests/smoke/conjugate_gradient.cpp]: `sqrt`->`std::sqrt` ?

File: tests/smoke/conjugate_gradient.cpp [hunk 18]
```diff
@@ -209,17 +242,24 @@ void grbProgram( const struct input &data_in, struct output &out ) {
 	set( x, static_cast< ScalarType >( 1 ) / static_cast< ScalarType >( n ) );
 	set( b, static_cast< ScalarType >( 1 ) );
```
Comment (anyzelman) [inline tests/smoke/conjugate_gradient.cpp]: `set`->`grb::set`?

File: tests/smoke/conjugate_gradient.cpp [hunk 19]
```diff
@@ -298,6 +338,14 @@ void grbProgram( const struct input &data_in, struct output &out ) {
 		return;
 	}
 
+	const double residual_norm = sqrt( out.residual );
```
Comment (anyzelman) [inline tests/smoke/conjugate_gradient.cpp]: std::sqrt

File: tests/smoke/conjugate_gradient.cpp [hunk 20]
```diff
@@ -298,6 +338,14 @@ void grbProgram( const struct input &data_in, struct output &out ) {
...
+	if( residual_norm >= tolerance ) {
+		// notify caller that the residual is too high
+		std::cerr << "Residual is too high: " << residual_norm << " >= " << tolerance << "\n";
+		out.error_code = 40;
+	}
```
Comment (anyzelman) [inline tests/smoke/conjugate_gradient.cpp]: In this case, the error code returned should also be `FAILED`

**Review by anyzelman — CHANGES_REQUESTED** [review]
This is great, only super minor comments (and it may needs rebasing-- I did see a force push from me yesterday, but unsure if i did that after the latest merge or not)

File: include/graphblas/algorithms/conjugate_gradient.hpp (hunk 2 above)
Comment (anyzelman) [reply include/graphblas/algorithms/conjugate_gradient.hpp:159]: Self-updating: there is one extremely compelling (though futuristic) use case-- that of a dispatching backend that given an incoming compute request, decides on which backend/device the computation should run, and then sends it there. Such a dispatcher was already envisioned on the core primitive level, but also make sense (perhaps even makes more sense) on the algorithm level. Made issue #282 to resolve this for all algorithms at a later stage.

TL;DR this comment can be ignored / resolved without any code changes

File: bootstrap.sh (hunk 1 above)
Comment (alberto-scolari) [reply bootstrap.sh]: solved in recent commit

File: include/graphblas/algorithms/conjugate_gradient.hpp (hunk 2 above)
Comment (alberto-scolari) [reply include/graphblas/algorithms/conjugate_gradient.hpp:159]: modification reverted

File: include/graphblas/algorithms/conjugate_gradient.hpp (hunk 3 above)
Comment (alberto-scolari) [reply include/graphblas/algorithms/conjugate_gradient.hpp]: modification reverted

File: include/graphblas/algorithms/conjugate_gradient.hpp (hunk 4 above)
Comment (alberto-scolari) [reply include/graphblas/algorithms/conjugate_gradient.hpp]: modification reverted

File: tests/smoke/conjugate_gradient.cpp (hunk 16 above)
Comment (alberto-scolari) [reply tests/smoke/conjugate_gradient.cpp]: modification reverted

File: tests/smoke/conjugate_gradient.cpp (hunk 17 above)
Comment (alberto-scolari) [reply tests/smoke/conjugate_gradient.cpp]: modification reverted; other occurrences fixed

File: tests/smoke/conjugate_gradient.cpp (hunk 18 above)
Comment (alberto-scolari) [reply tests/smoke/conjugate_gradient.cpp]: modification reverted

File: tests/smoke/conjugate_gradient.cpp (hunk 19 above)
Comment (alberto-scolari) [reply tests/smoke/conjugate_gradient.cpp]: modification reverted

File: tests/smoke/conjugate_gradient.cpp (hunk 20 above)
Comment (alberto-scolari) [reply tests/smoke/conjugate_gradient.cpp]: modification reverted

File: include/graphblas/nonblocking/vector.hpp (hunk 5 above)
Comment (alberto-scolari) [reply include/graphblas/nonblocking/vector.hpp:199]: fixed

File: include/transition/kml_iss.h (hunk 6 above)
Comment (alberto-scolari) [reply include/transition/kml_iss.h:145]: fixed

File: include/transition/kml_iss.h (hunk 7 above)
Comment (alberto-scolari) [reply include/transition/kml_iss.h]: fixed

File: include/transition/kml_iss.h (hunk 8 above)
Comment (alberto-scolari) [reply include/transition/kml_iss.h]: fixed

File: src/transition/CMakeLists.txt (hunk 9 above)
Comment (alberto-scolari) [reply src/transition/CMakeLists.txt]: fixed naming and disabling reference[_omp]-based targets by default; optionally enabled via CMake or bootstrap script

File: src/transition/CMakeLists.txt (hunk 10 above)
Comment (alberto-scolari) [reply src/transition/CMakeLists.txt]: fixed

File: src/transition/kml_iss.cpp (hunk 11 above)
Comment (alberto-scolari) [reply src/transition/kml_iss.cpp]: fixed

File: src/transition/kml_iss.cpp (hunk 13 above)
Comment (alberto-scolari) [reply src/transition/kml_iss.cpp]: fixed

**Comment by anyzelman** [comment]
Manually running full test suite with LPF, and waiting GitHub and internal CIs after final code review mods of [8bd4707](https://github.com/Algebraic-Programming/ALP/pull/281/commits/8bd4707ed50f803f3cc538f642075d7f30e55056)

**Comment by anyzelman** [comment]
@alberto-scolari I tested the spblas prefix option but it seems to have no effect: the header and library for the KML solver are still called kml_, and the emitted prefix macro is unchanged as well. I'm also wondering about what prefix should be changed-- should the kml solver interface not always remain the same, and spblas should instead change the spblas.h only? (The doc at bootstrap.sh says only the kml prefix would change.)

The blas_sparse.h header I believe should never change prefixes, as that is the only properly standardised api (as by Duff et al.). At most the prefixes of the extended sparse_blas API (currently prefixed EXTBLAS_) could be configurable.

The spblas API is like an industry de-facto type standard where in fact it is common to allow for changed prefixes, so the option could well apply there.

The MKL solver API is non-standard afaik so can be kept without a capability of changing its prefix. Its implementation in relying on solver.h is really nice btw.

In summary,
1. the --spblas-prefix option should modify the function names in spblas.h only, and leave the mkl stuff unchanged (both its file names and function names);
2. the bootstrap.sh doc should be updated such (in that the option changes function names in spblas.h only);
3. (and primarily:) currently the option seems to do nothing and should be fixed (or removed and kept for a separate MR-- maybe easier)

I meanwhile rebased and relaunched all tests, this time also on ARM for a change with recent tests runs:)

**Comment by anyzelman** [comment]
At this point, all manual tests (unit+smoke) on ARM with LPF pass; all CIs also pass.

**Comment by anyzelman** [comment]
Many thanks @alberto-scolari -- it sadly did however reset a lot of my previous review so please allow me to re-review and merge it tomorrow;) I'll mean time rebase and run the test suite.

**Review by anyzelman — DISMISSED** [review]

**Comment by anyzelman** [comment]
Review OK, manual smoke test with LPF OK. Waiting CIs.

**Comment by anyzelman** [comment]
All CIs OK

**Review by anyzelman — APPROVED** [review]
